src/lavague/evaluator.py

```diff
@@ -0,0 +1,245 @@
+import pandas as pd
+import re
+from tqdm import tqdm
+
+decontaminate_html = lambda x: re.sub(r' backend_node_id="\d+"', '', x)
+
+def keep_assignments(code_snippet):
+    # Regex to match variable assignments. This pattern assumes variable names are valid Python identifiers
+    # and captures typical assignment statements, excluding those that might appear in comments or strings.
+    pattern = r'^\s*[a-zA-Z_][a-zA-Z0-9_]*\s*=\s*.+'
+
+    # Filter and keep only lines with variable assignments
+    assignments = [line for line in code_snippet.split('\n') if re.match(pattern, line)]
+
+    # Join the filtered lines back into a string
+    return "\n".join(assignments)
+
+# This function will be used to visit each node in the AST
+class VariableVisitor(ast.NodeVisitor):
+    def __init__(self):
+        super().__init__()
+        self.output = []
+    
+    def visit_Assign(self, node):
+        
+        # For each assignment, print the targets (variables)
+        for target in node.targets:
+            if isinstance(target, ast.Name):  # Ensure it's a variable assignment
+                self.output.append(target.id)
+
+def return_first_assignment_variables(code_snippet):
+    # Parse the code snippet into an abstract syntax tree (AST)
+    parsed = ast.parse(code_snippet)
+    
+    # Create a VariableVisitor object
+    visitor = VariableVisitor()
+    
+    # Visit (i.e., traverse) the AST nodes
+    visitor.visit(parsed)
+    
+    # Return the variables found
+    return visitor.output[0]
+
+def inject_backend_node_id(html):
+    # Parse the HTML content
+    soup = BeautifulSoup(html, 'html.parser')
+    
+    # Initialize a unique ID counter
+    backend_node_id = 1
+    
+    # Loop through each element in the HTML
+    for element in soup.find_all(True):  # True finds all tags
+        # Add the 'backend_node_id' attribute with the current ID
+        element['backend_node_id'] = backend_node_id
+        # Increment the ID for the next element
+        backend_node_id += 1
+    
+    # Return the modified HTML as a string
+    return str(soup)
+
+def contains_backend_node_id(html_content):
+    # Parse the HTML content
+    soup = BeautifulSoup(html_content, 'html.parser')
+    
+    # Search for any tags with the 'backend_node_id' attribute
+    if soup.find(attrs={"backend_node_id": True}):
+        # If any such tags are found, return True
+        return True
+    # If no such tags are found, return False
+    return False
+
+def extract_backend_node_ids(html_content):
+    soup = BeautifulSoup(html_content, 'html.parser')
+    return set([tag['backend_node_id'] for tag in soup.find_all(attrs={"backend_node_id": True})])
+
+def id_recall(ground_truth_outer_html, retrieved_context):
+    ground_truth_ids = extract_backend_node_ids(ground_truth_outer_html)
+    context_ids = extract_backend_node_ids(retrieved_context)
+    recall = len(ground_truth_ids & context_ids) / len(ground_truth_ids)
```

Comment:
For some reason, when i am running locally
len(ground_truth_ids) is 0 and throwing `ZeroDivisionError: division by zero`